### ISSUE TITLE
[FEATURE] Improve the interactivity by using mouse and keyboard

### ISSUE THREAD

**Issue by xareelee** [bokeh/bokeh#13093]
### Problem description

I just read the tutorial from holoviz. In chapter 2, when [using `widget` mode for matplotlib](https://holoviz.org/tutorial/Plotting.html#using-pandas-plot), you can use left click button to pan and use right click button to zoom, where bokeh can't.

It's very inefficient to review and study many plots while you need to switch the control mode in the toolbar. For example, if you want to pan first and zoom in, and then pan, and the zoom in... to focus on specific data, you need to click the toolbar multiple times to switch between pan/zoom mode for mouse left click.

### Feature description

What if:

- When the mouse cursor is **hover over the plot** and **holding `shift`**, 
    - you can **zoom in/out** by **mouse wheel** (zoom-in center is where your mouse cursor is, just like Google Maps).
    - you can **select a box to zoom in** by **mouse left clicking and dragging**.
    - you can **undo** previous action by **mouse right clicking** (you can right click multiple times till the plot is reset to original view).
- When the mouse cursor is **hover over the plot** and **without holding `shift`**, 
    - you can **pan the plot** by **mouse left clicking and dragging**.
    - you can **zoom in and center the plot where you choose** by **DOUBLE mouse left clicking** (just like Google Maps).
    - scrolling the webpage by just **mouse wheel** as default.

### Potential alternatives

For non-wheel mouses or preferring using keyboard, you also can use the keystroke to zoom in/out and pan:

- When the mouse cursor is **hover over the plot** and **without holding `shift`**, 
    - you can **pan the plot** by keystroke **WASD** (up/left/down/right) about moving 25%~33% of plotting view size for each keystroke.
- When the mouse cursor is **hover over the plot** and **holding `shift`**, 
    - you can **pan the plot faster** by keystroke **WASD** (up/left/down/right) about moving 66% of plotting view size for each keystroke.
    - you can **zoom in/out** by keystroke **E/Q** (zoom-in center is where your mouse cursor is, just like Google Maps).
    - you can **undo** previous action by **R** (you can do this multiple times till the plot is reset to original view).

**Comment by bryevdv** [comment]
@xareelee you are aware that Bokeh can have the drag tool and the wheel zoom tool  active at same time, and you can left-drag to pan and scroll (no shift required!) to zoom in and out without any switching? I also happen to think the scroll zoom is far superior for this exploration, since you can position the cursor directly over what you want to zoom in. 

Offhand I am pretty 👎 on the idea of adding keyboard bindings. There's just too many ways that can potentially conflict with other widgets or non-Bokeh elements, and the conditional aspect on focus that can't really easily conveyed, seems like an invitation for confusion.

Regarding right-drag, it just seems highly unusual. I've never encountered it in any UI that I can think of. Are there any other examples of this besides MPL?

cc @bokeh/dev

**Comment by xareelee** [comment]
@bryevdv When you scroll the web page, if your cursor hovers over the plot, it may trigger unwanted zoom in/out while scrolling the web page. Besides, you need to enable the scrolling feature by clicking the button on the tool bar first. If you need to review lots of data/plots daily or hourly, it would be very annoying.

With `shift` binding, you can intentionally scroll to zoom in/out, pick an area to zoom in, and right click to undo. You don't need to focus on the toolbar; just focus on the plot.

**Comment by water-e** [comment]
Hi all, is the solution already settled? I've used mouse-keyboard combinations and whereas I don't think I would mind any of the proposals I don't think it is the "killer app" that is missing from matplotlib. That prize goes to the right click drag mouse gesture in the matplotlib pan tool. It stretches or compresses the view limits on one axis or the other (intelligently with respect to direction). Done fluidly, and coupled with a zoom stack, you don't need the zoom any more for interactive work, perhaps for fine tuning. I can't remember how I felt about it in the early days when I discovered it and whether it would cause false-gestures to a new user, but to an experienced user it never causes problems. Honestly, it is a tad hard to explain, but the lack of it has kept me from ever fully embracing bokeh for production work, and having seen folks mousing and clicking around in my organization I feel it is like they don't even know how fluid interactive graphics can be (no great love for baggage and limitations that go with it, which is why I'm here).
